Fix declared type of custom comparer for MaxBy and MinBy

IDisposable opened the pull request:
Use the `TKey` output of the keySelector (not the input `TSource`) for the the custom `IComparer` since that's what the keySelector will emit.

Fixes #113878

This is a semi-**BREAKING CHANGE** but is justifiable in that the only code that could have worked before was when `TKey` and `TSource` were the same type (e.g. both `int` as in the [test](https://github.com/dotnet/runtime/pull/113944/files#diff-772e85f589f6ca72a3c1219cd1927051aee02146512ce99395eb7ce476ea45afR635-R640) [cases](https://github.com/dotnet/runtime/pull/113944/files#diff-92bc6c71e60aa54608593d0c217ecdc98d8fdbdab3957ac045f29281381dc96dR603-R608))... and that code will continue to compile correctly.

~This includes an [APICompat update](https://learn.microsoft.com/en-us/dotnet/fundamentals/apicompat/overview) with `/p:ApiCompatGenerateSuppressionFile=true`~

IDisposable committed:
Fix declared type of custom comparer for MaxBy and MinBy

Use the `TKey` output of the keySelector (not the input `TSource`) for the the custom `IComparer` since that's what the keySelector will emit.

Fixes #113878

This is a semi-**BREAKING CHANGE** but is justifiable in that the only code that could have worked before was when `TKey` and `TSource` were the same type (e.g. both `int` as in the test cases)... and that code will continue to compile correctly.

IDisposable commented:
> So if `TSource` is the same as `TKey`, then this pull request does not break source compatibility. It still breaks binary compatibility, though.

Seeking clarification, it certainly doesn't change runtime-behavior in this case (e.g. it used to work, and still does). It does change the declaration and thus the method signature, but no compile-time difference would be visible, correct?

IDisposable commented:
> Although different types in `MinBy`/`MaxBy` did not work with `EnumerableQuery<T>`, some other `IQueryProvider` may have supported them.

Seeking clarification again, these are `static` extension methods declared against static class `Queryable`, and the code body does a `Expression.Call` using the "specified" method using (e.g. for MinBy)
```c#
new Func<IQueryable<TSource>, Expression<Func<TSource, TKey>>, TSource?>(MinBy).Method
```

This is compile-time bound to the implementation in _/src/libraries/System.Linq/src/System/Linq/Min.cs_
```c#
public static TSource? MinBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IComparer<TKey>? comparer)
```

So how does another `IQueryProvider` get inserted into this?

I know I'm missing something...

IDisposable commented:
Just let me know if I have to go back and mark the old ones obsolete!  I was just proceeding based on [this](https://github.com/dotnet/runtime/issues/113878#issuecomment-2751454244) but would be happy to insert back the old ones :)

> It seems that we might just afford to make a breaking change, but if not we could just mark the incorrect overload as obsolete/`EB.Never`/`OverloadResolutionPriority(int.MinValue)`.

IDisposable commented:
@KalleOlaviNiemitalo thanks tons for the explainer... I'm a bit rusty on IL resolution games.

eiriktsarpalis commented:
Let's hold off from merging until a final decision has been made in API review.

eiriktsarpalis commented:
Removing APIs introduces a lot of complications which would rather avoid. Even though I'm still waiting on approval the direction we're going for is adding the correct APIs while obsolete the incorrect ones:
```diff
public static class Queryable
{
+   [Obsolete("SYSLIBxxxx"), EditorBrowsable(EditorBrowsableState.Never), OverloadResolutionPriority(-1)]
    public static TSource? MinBy<TSource, TKey>(this IQueryable<TSource> source, Expression<Func<TSource, TKey>> keySelector, IComparer<TSource>? comparer);
+   [Obsolete("SYSLIBxxxx"), EditorBrowsable(EditorBrowsableState.Never), OverloadResolutionPriority(-1)]
    public static TSource? MaxBy<TSource, TKey>(this IQueryable<TSource> source, Expression<Func<TSource, TKey>> keySelector, IComparer<TSource>? comparer);
+   public static TSource? MinBy<TSource, TKey>(this IQueryable<TSource> source, Expression<Func<TSource, TKey>> keySelector, IComparer<TKey>? comparer);
+   public static TSource? MaxBy<TSource, TKey>(this IQueryable<TSource> source, Expression<Func<TSource, TKey>> keySelector, IComparer<TKey>? comparer);
}
```
We have an [obsoletion process](https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md#obsoletions) in place which involves picking the next available SYSLIB id from the list of obsoletions in the linked document.

eiriktsarpalis commented:
The above has now been approved as proposed. Please update the PR as convenient.

IDisposable committed:
Update to keep the old MinBy and MaxBy around as Obsolete

Ensures the binary and source compatibility in case anyone was using the broken overloads.

IDisposable commented:
Obsoletions added (not sure if this will require some APICompat, but reverted my changes in that XML file.

IDisposable commented on docs/project/list-of-diagnostics.md:
```diff
@@ -115,6 +115,7 @@ The PR that reveals the implementation of the `<IncludeInternalObsoleteAttribute
 |  __`SYSLIB0058`__ | KeyExchangeAlgorithm, KeyExchangeStrength, CipherAlgorithm, CipherStrength, HashAlgorithm and HashStrength properties of SslStream are obsolete. Use NegotiatedCipherSuite instead. |
 |  __`SYSLIB0059`__ | SystemEvents.EventsThreadShutdown callbacks are not run before the process exits. Use AppDomain.ProcessExit instead. |
 |  __`SYSLIB0060`__ | The constructors on Rfc2898DeriveBytes are obsolete. Use the static Pbkdf2 method instead. |
+|  __`SYSLIB0061`__ | The Queryable MinBy and MaxBy taking an IComparer<TSource> are obsolete. Use the new ones that take an IComparer<TKey>. |
```
You are correct. I'll fix that, thanks!

There's also some unescaped `<` (though in single quotes `'`) down at [SYSLIB1062](https://github.com/dotnet/runtime/pull/113944/files/aac10a3012daba99da86fd3c4d67654a309f8348#diff-a390b1ffbc64ee3385b5b5b8c52cbf70851625db7dd42f4c666e336bc22c8a3bR189) and they don't render right. 

```
Project must be updated with '<AllowUnsafeBlocks>true</AllowUnsafeBlocks>'
```
looks like 
![image](https://github.com/user-attachments/assets/2ff8e6ec-94ed-4877-8340-9b4c857d127e)


Should I fix those too?

IDisposable replied:
Went ahead and updated both places
![image](https://github.com/user-attachments/assets/fffa0e3a-c691-49f3-9bbb-e521bb4c53fb)
![image](https://github.com/user-attachments/assets/7f95cedc-5448-49e7-855d-9b0d9e45f75c)

IDisposable committed:
PR Feedback: Fixed missing markdown escaping for < and >

IDisposable commented on src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs:
```diff
@@ -1780,11 +1782,14 @@ public static long LongCount<TSource>(this IQueryable<TSource> source, Expressio
         /// <exception cref="ArgumentNullException"><paramref name="source" /> is <see langword="null" />.</exception>
-        /// <exception cref="ArgumentException">No key extracted from <paramref name="source" /> implements the <see cref="IComparable" /> or <see cref="IComparable{TKey}" /> interface.</exception>
+        /// <exception cref="ArgumentException">No key extracted from <paramref name="source" /> implements the <see cref="IComparable" /> or <see cref="IComparable{TSource}" /> interface.</exception>
         [DynamicDependency("MinBy`2", typeof(Enumerable))]
```
Agreed, I didn't see a reason not to duplicate it in when making the new ones, nor stripping from the old ones.

eiriktsarpalis reviewed (approved):
Excellent. Thank you for the contribution.

hez2010 commented:
Should this be backported? Otherwise those APIs are not usable in earlier versions of .NET.

eiriktsarpalis commented:
I don't think this meets the bar for servicing.

IDisposable commented:
Breaking change documentation in docs PR https://github.com/dotnet/docs/pull/45565

huoyaoyuan commented:
Shouldn't the NO-MERGE label be removed before the merge? Or haven't we applied rule for this?

eiriktsarpalis commented:
There's no rule enforcing this.

ericstj commented:
Removing `needs-breaking-change-doc-created` label as this PR already has a documentation issue: https://github.com/dotnet/docs/issues/45535
